packages/db/src/collection.ts

```diff
@@ -430,63 +300,380 @@ export class Collection<T extends object = Record<string, unknown>> {
+    for (const event of events) {
+      this.emitEvent(event)
+    }
+
+    // Convert to ChangeMessage format and emit to subscribeChanges listeners
+    if (events.length > 0) {
+      const changeMessages: Array<ChangeMessage<T>> = events.map((event) => {
+        const changeMessage: ChangeMessage<T> = {
+          type: event.type,
+          key: event.key,
+          value: event.value,
+        }
+
+        if (event.previousValue) {
+          ;(changeMessage as any).previousValue = event.previousValue
+        }
+
+        return changeMessage
+      })
+
+      for (const listener of this.changesBatchListeners) {
+        listener([...changeMessages])
+      }
+    }
+  }
+
+  /**
+   * Emit an event to individual listeners (not batched)
+   */
+  private emitEvent(event: CollectionEvent<T>): void {
+    // Emit to general listeners
+    for (const listener of this.eventListeners) {
+      listener(event)
+    }
+
+    // Emit to key-specific listeners
+    const keyListeners = this.keyListeners.get(event.key)
+    if (keyListeners) {
+      for (const listener of keyListeners) {
+        listener(
+          event.type === `delete` ? undefined : event.value,
+          event.previousValue
+        )
+      }
+    }
+  }
+
+  /**
+   * Subscribe to collection events
+   */
+  public subscribe(listener: EventListener<T>): () => void {
+    this.eventListeners.add(listener)
+    return () => {
+      this.eventListeners.delete(listener)
+    }
+  }
+
+  /**
+   * Subscribe to changes for a specific key
+   */
+  public subscribeKey(key: string, listener: KeyListener<T>): () => void {
+    if (!this.keyListeners.has(key)) {
+      this.keyListeners.set(key, new Set())
+    }
+    this.keyListeners.get(key)!.add(listener)
+
+    return () => {
+      const listeners = this.keyListeners.get(key)
+      if (listeners) {
+        listeners.delete(listener)
+        if (listeners.size === 0) {
+          this.keyListeners.delete(key)
+        }
+      }
+    }
+  }
+
+  /**
+   * Get the current value for a key (virtual derived state)
+   */
+  public get(key: string): T | undefined {
+    // Check if optimistically deleted
+    if (this.derivedDeletes.has(key)) {
+      return undefined
+    }
+
+    // Check optimistic upserts first
+    if (this.derivedUpserts.has(key)) {
+      return this.derivedUpserts.get(key)
+    }
+
+    // Fall back to synced data
+    return this.syncedData.get(key)
+  }
+
+  /**
+   * Check if a key exists in the collection (virtual derived state)
+   */
+  public has(key: string): boolean {
+    // Check if optimistically deleted
+    if (this.derivedDeletes.has(key)) {
+      return false
+    }
+
+    // Check optimistic upserts first
+    if (this.derivedUpserts.has(key)) {
+      return true
+    }
+
+    // Fall back to synced data
+    return this.syncedData.has(key)
+  }
+
+  /**
+   * Get the current size of the collection (virtual)
+   */
+  public get size(): number {
```

Comment:
recompute this incrementally in `recomputeOptimisticState` & when data is synced